File: frontend/src/utils/excel/dealerMetadataService.js
import { readWorkbook, sheetTo2DArray, cleanStr } from './excelUtils.js';

const README_SHEET_NAMES = new Set(['read me', 'readme', 'read_me']);

const FIELD_ALIASES = {
  gstin: ['gstin', "taxpayer's gstin", 'taxpayers gstin', 'taxpayer gstin'],
  legal_name: ['legal name', 'legal name of taxpayer', 'legalname'],
  trade_name: ['trade name', 'trade name (if any)', 'tradename'],
  financial_year: ['financial year', 'financialyear', 'fy'],
  tax_period: ['tax period', 'tax period ', 'taxperiod'],
  arn: ['arn'],
  arn_date: ['arn date'],
  download_date: [
    'date and time of generation',
    'date of generation',
    'download date',
    'date of download',
  ],
};

const MONTH_NAMES = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

const MONTH_NAME_MAP = {
  january: 1, jan: 1, '01': 1, '1': 1,
  february: 2, feb: 2, '02': 2, '2': 2,
  march: 3, mar: 3, '03': 3, '3': 3,
  april: 4, apr: 4, '04': 4, '4': 4,
  may: 5, '05': 5, '5': 5,
  june: 6, jun: 6, '06': 6, '6': 6,
  july: 7, jul: 7, '07': 7, '7': 7,
  august: 8, aug: 8, '08': 8, '8': 8,
  september: 9, sep: 9, sept: 9, '09': 9, '9': 9,
  october: 10, oct: 10, '10': 10,
  november: 11, nov: 11, '11': 11,
  december: 12, dec: 12, '12': 12,
};

const GSTR1_FALLBACK = {
  financial_year: [3, 2], // 0-based row 3 (Excel 4), col 2 (Excel C)
  tax_period: [4, 2],
  gstin: [5, 2],
  legal_name: [6, 2],
  trade_name: [7, 2],
  arn: [8, 2],
  arn_date: [9, 2],
  download_date: [10, 2],
};

const GSTR2A_FALLBACK = {
  gstin: [1, 2], // row 2, col C
  legal_name: [2, 2],
  trade_name: [3, 2],
  tax_period: [1, 4], // row 2, col E
  financial_year: [2, 4], // row 3, col E
  download_date: [3, 4],
};

export function normalizeMetadataLabel(text) {
  if (text == null) return '';
  return String(text).trim().toLowerCase().replace(/\s+/g, ' ');
}

export function matchMetadataField(label) {
  const normalized = normalizeMetadataLabel(label);
  if (!normalized) return null;
  for (const [field, aliases] of Object.entries(FIELD_ALIASES)) {
    for (const alias of aliases) {
      if (normalized === alias || normalized.includes(alias) || alias.includes(normalized)) {
        return field;
      }
    }
  }
  return null;
}

/**
 * Parses financial year string like '2022-23' or '2022-2023' into { startYear, endYear, raw }
 */
export function parseFinancialYear(rawFy) {
  const text = cleanStr(rawFy);
  if (!text) return { startYear: null, endYear: null, raw: '' };
  
  const m = text.match(/^(\d{4})[-/](\d{2,4})$/);
  if (m) {
    const startYear = parseInt(m[1], 10);
    let endYear = parseInt(m[2], 10);
    if (endYear < 100) {
      endYear = Math.floor(startYear / 100) * 100 + endYear;
    }
    const shortEnd = String(endYear).slice(-2);
    return {
      startYear,
      endYear,
      raw: `${startYear}-${shortEnd}`,
    };
  }
  return { startYear: null, endYear: null, raw: text };
}

/**
 * Parses GSTR-2A Tax Period (MMYYYY or numeric like 042022, 12023) into { month, year, display }
 */
export function parseGstr2aTaxPeriod(rawPeriod, financialYear = '') {
  if (rawPeriod == null) return { month: null, year: null, display: '', raw: '' };
  let str = String(rawPeriod).trim();
  if (!str) return { month: null, year: null, display: '', raw: '' };

  // Handle numeric like 12023 (5 digits) -> '012023'
  if (/^\d{5}$/.test(str)) {
    str = '0' + str;
  }

  // MMYYYY
  if (/^\d{6}$/.test(str)) {
    const month = parseInt(str.slice(0, 2), 10);
    const year = parseInt(str.slice(2), 10);
    if (month >= 1 && month <= 12) {
      const monthName = MONTH_NAMES[month - 1];
      return {
        month,
        year,
        display: `${monthName}-${year}`,
        raw: str,
      };
    }
  }

  // If already month name, delegate to GSTR-1 style
  return parseGstr1TaxPeriod(rawPeriod, financialYear);
}

/**
 * Parses GSTR-1 Tax Period (e.g. 'January', 'April') using Financial Year to resolve the year
 */
export function parseGstr1TaxPeriod(rawPeriod, financialYear = '') {
  if (rawPeriod == null) return { month: null, year: null, display: '', raw: '' };
  const str = String(rawPeriod).trim();
  if (!str) return { month: null, year: null, display: '', raw: '' };

  // If MMYYYY format was passed in GSTR-1
  if (/^\d{5,6}$/.test(str)) {
    return parseGstr2aTaxPeriod(str, financialYear);
  }

  const cleanMonthKey = str.toLowerCase().replace(/[^a-z0-9]/g, '');
  const month = MONTH_NAME_MAP[cleanMonthKey];

  if (month) {
    const monthName = MONTH_NAMES[month - 1];
    const fyInfo = parseFinancialYear(financialYear);
    let year = null;

    if (fyInfo.startYear && fyInfo.endYear) {
      // April (4) to December (12) belongs to startYear; January (1) to March (3) belongs to endYear
      year = month >= 4 ? fyInfo.startYear : fyInfo.endYear;
    }

    const display = year ? `${monthName}-${year}` : monthName;
    return {
      month,
      year,
      display,
      raw: str,
    };
  }

  // If format is 'Month-YYYY' or 'Month YYYY'
  const myMatch = str.match(/^([A-Za-z]+)[-\s](\d{4})$/);
  if (myMatch) {
    const mKey = myMatch[1].toLowerCase();
    const m = MONTH_NAME_MAP[mKey];
    const y = parseInt(myMatch[2], 10);
    if (m) {
      return {
        month: m,
        year: y,
        display: `${MONTH_NAMES[m - 1]}-${y}`,
        raw: str,
      };
    }
  }

  return { month: null, year: null, display: str, raw: str };
}

/**
 * Build chronological range display from an array of normalized period objects:
 * e.g. 'April-2022 to March-2023' or 'April-2022 to June-2022' or 'April-2022'
 */
export function buildTaxPeriodRangeDisplay(periodObjects) {
  const validPeriods = periodObjects.filter((p) => p && p.month && p.year);
  if (validPeriods.length === 0) {
    const rawDisplays = periodObjects.map((p) => p?.display || p?.raw).filter(Boolean);
    return rawDisplays[0] || '';
  }

  // Sort chronologically by year then month
  const sorted = [...validPeriods].sort((a, b) => {
    if (a.year !== b.year) return a.year - b.year;
    return a.month - b.month;
  });

  const first = sorted[0];
  const last = sorted[sorted.length - 1];

  if (first.year === last.year && first.month === last.month) {
    return first.display;
  }

  return `${first.display} to ${last.display}`;
}

export function extractDealerMetadataFromSheet(rows, returnType) {
  const found = {};
  const maxRows = Math.min(rows.length, 25);

  for (let r = 0; r < maxRows; r++) {
    const row = rows[r] || [];
    const maxCols = Math.min(row.length, 10);
    for (let c = 0; c < maxCols; c++) {
      const cellVal = row[c];
      const field = matchMetadataField(cellVal);
      if (!field || found[field]) continue;

      // Scan right for value
      for (let vc = c + 1; vc < maxCols; vc++) {
        const val = cleanStr(row[vc]);
        if (!val) continue;
        if (matchMetadataField(val)) break;
        found[field] = val;
        break;
      }
    }
  }

  // Apply positional fallbacks if missing
  const fallback = returnType === 'gstr1' ? GSTR1_FALLBACK : GSTR2A_FALLBACK;
  for (const [field, [r, c]] of Object.entries(fallback)) {
    if (!found[field] && rows[r] && rows[r][c]) {
      const val = cleanStr(rows[r][c]);
      if (val) found[field] = val;
    }
  }

  const gstin = cleanStr(found.gstin).toUpperCase();
  const legalName = cleanStr(found.legal_name);
  const tradeName = cleanStr(found.trade_name);
  const fyInfo = parseFinancialYear(found.financial_year);
  const financialYear = fyInfo.raw;

  const periodInfo =
    returnType === 'gstr1'
      ? parseGstr1TaxPeriod(found.tax_period, financialYear)
      : parseGstr2aTaxPeriod(found.tax_period, financialYear);

  return {
    dealer_id: gstin ? `dealer_${gstin}` : '',
    gstin,
    legal_name: legalName,
    trade_name: tradeName,
    financial_year: financialYear,
    tax_period: periodInfo.display || cleanStr(found.tax_period),
    tax_period_raw: periodInfo.raw,
    tax_period_month: periodInfo.month,
    tax_period_year: periodInfo.year,
    tax_period_display: periodInfo.display,
    // CamelCase aliases
    legalName,
    tradeName,
    financialYear,
    taxPeriod: periodInfo.display || cleanStr(found.tax_period),
    taxPeriodRaw: periodInfo.raw,
    taxPeriodMonth: periodInfo.month,
    taxPeriodYear: periodInfo.year,
    taxPeriodDisplay: periodInfo.display,
    // Other fields
    arn: cleanStr(found.arn),
    arn_date: cleanStr(found.arn_date),
    download_date: cleanStr(found.download_date),
  };
}

export async function extractDealerMetadataFromFile(file, returnType) {
  const wb = await readWorkbook(file);
  const readmeName = wb.SheetNames.find((name) => README_SHEET_NAMES.has(name.trim().toLowerCase()));
  if (!readmeName) {
    return {
      dealer_id: '',
      gstin: '',
      legal_name: '',
      trade_name: '',
      financial_year: '',
      tax_period: '',
      tax_period_raw: '',
      tax_period_month: null,
      tax_period_year: null,
      tax_period_display: '',
      legalName: '',
      tradeName: '',
      financialYear: '',
      taxPeriod: '',
      taxPeriodRaw: '',
      taxPeriodMonth: null,
      taxPeriodYear: null,
      taxPeriodDisplay: '',
      arn: '',
      arn_date: '',
      download_date: '',
    };
  }

  const sheet = wb.Sheets[readmeName];
  const rows = sheetTo2DArray(sheet);
  return extractDealerMetadataFromSheet(rows, returnType);
}

export async function extractDealerMetadataFromFiles(files, returnType) {
  const results = [];
  for (const file of files) {
    const dealer = await extractDealerMetadataFromFile(file, returnType);
    results.push({ filename: file.name, dealer });
  }

  // Validate consistency across files
  const first = results[0]?.dealer;
  const mismatches = [];

  if (first?.gstin) {
    for (const item of results) {
      if (item.dealer.gstin && item.dealer.gstin !== first.gstin) {
        mismatches.push({
          field: 'GSTIN',
          source_file: item.filename,
          expected: first.gstin,
          found: item.dealer.gstin,
        });
      }
    }
  }

  if (first?.financial_year) {
    for (const item of results) {
      if (item.dealer.financial_year && item.dealer.financial_year !== first.financial_year) {
        mismatches.push({
          field: 'Financial Year',
          source_file: item.filename,
          expected: first.financial_year,
          found: item.dealer.financial_year,
        });
      }
    }
  }

  if (mismatches.length > 0) {
    const err = new Error('Dealer metadata mismatch across files');
    err.payload = {
      error_type: 'dealer_mismatch',
      mismatches,
      message: mismatches
        .map(
          (m) =>
            `${m.field} mismatch in ${m.source_file}: expected "${m.expected}", found "${m.found}"`,
        )
        .join('; '),
    };
    throw err;
  }

  // Compute period range from all selected valid files
  const periodObjects = results.map((r) => ({
    month: r.dealer.tax_period_month,
    year: r.dealer.tax_period_year,
    display: r.dealer.tax_period_display,
    raw: r.dealer.tax_period_raw,
  }));
  const periodRangeDisplay = buildTaxPeriodRangeDisplay(periodObjects);

  const combinedDealer = {
    ...(first || {}),
    tax_period: periodRangeDisplay || first?.tax_period || '',
    taxPeriod: periodRangeDisplay || first?.taxPeriod || '',
  };

  return {
    dealer: combinedDealer,
    files_analyzed: results.length,
    file_metadata: results,
    workbook_id: `wb_${returnType}_${first?.gstin || 'unknown'}_${Date.now()}`,
    return_type: returnType,
    source_files: files.map((f) => f.name),
  };
}

File: backend/services/dealer_metadata_service.py
"""Extract dealer metadata from GSTR Read me worksheets."""

from __future__ import annotations

import io
import re
from typing import Dict, List, Optional, Tuple, Any

import openpyxl

from models.dealer_metadata import DealerMetadata

README_SHEET_NAMES = {"read me", "readme", "read_me"}

FIELD_ALIASES: Dict[str, Tuple[str, ...]] = {
    "gstin": ("gstin", "taxpayer's gstin", "taxpayers gstin", "taxpayer gstin"),
    "legal_name": ("legal name", "legal name of taxpayer", "legalname"),
    "trade_name": ("trade name", "trade name (if any)", "tradename"),
    "financial_year": ("financial year", "financialyear", "fy"),
    "tax_period": ("tax period", "tax period ", "taxperiod"),
    "arn": ("arn",),
    "arn_date": ("arn date",),
    "download_date": (
        "date and time of generation",
        "date of generation",
        "download date",
        "date of download",
    ),
}

MONTH_NAMES = [
    "January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December"
]

MONTH_NAME_MAP = {
    "january": 1, "jan": 1, "01": 1, "1": 1,
    "february": 2, "feb": 2, "02": 2, "2": 2,
    "march": 3, "mar": 3, "03": 3, "3": 3,
    "april": 4, "apr": 4, "04": 4, "4": 4,
    "may": 5, "05": 5, "5": 5,
    "june": 6, "jun": 6, "06": 6, "6": 6,
    "july": 7, "jul": 7, "07": 7, "7": 7,
    "august": 8, "aug": 8, "08": 8, "8": 8,
    "september": 9, "sep": 9, "sept": 9, "09": 9, "9": 9,
    "october": 10, "oct": 10, "10": 10,
    "november": 11, "nov": 11, "11": 11,
    "december": 12, "dec": 12, "12": 12,
}

# Known GST portal cell positions (1-based row, col) when label scan fails.
GSTR1_FALLBACK: Dict[str, Tuple[int, int]] = {
    "financial_year": (4, 3),
    "tax_period": (5, 3),
    "gstin": (6, 3),
    "legal_name": (7, 3),
    "trade_name": (8, 3),
    "arn": (9, 3),
    "arn_date": (10, 3),
    "download_date": (11, 3),
}

GSTR2A_FALLBACK: Dict[str, Tuple[int, int]] = {
    "gstin": (2, 3),
    "legal_name": (3, 3),
    "trade_name": (4, 3),
    "tax_period": (2, 5),
    "financial_year": (3, 5),
    "download_date": (4, 5),
    "arn": (0, 0),
    "arn_date": (0, 0),
}


def _normalize_label(text: object) -> str:
    if text is None:
        return ""
    return re.sub(r"\s+", " ", str(text).strip().lower())


def _clean_value(value: object) -> str:
    if value is None:
        return ""
    return str(value).strip()


def find_readme_sheet_name(sheet_names: List[str]) -> Optional[str]:
    for name in sheet_names:
        if name.strip().lower() in README_SHEET_NAMES:
            return name
    return None


def _match_field(label: str) -> Optional[str]:
    normalized = _normalize_label(label)
    if not normalized:
        return None
    for field, aliases in FIELD_ALIASES.items():
        if normalized in aliases:
            return field
        for alias in aliases:
            if alias in normalized or normalized in alias:
                return field
    return None


def parse_financial_year(raw_fy: object) -> Tuple[Optional[int], Optional[int], str]:
    text = _clean_value(raw_fy)
    if not text:
        return None, None, ""
    m = re.match(r"^(\d{4})[-/](\d{2,4})$", text)
    if m:
        start_year = int(m.group(1))
        end_year = int(m.group(2))
        if end_year < 100:
            end_year = (start_year // 100) * 100 + end_year
        short_end = str(end_year)[-2:]
        return start_year, end_year, f"{start_year}-{short_end}"
    return None, None, text


def parse_gstr2a_tax_period(raw_period: object, financial_year: str = "") -> Tuple[Optional[int], Optional[int], str, str]:
    if raw_period is None:
        return None, None, "", ""
    s = str(raw_period).strip()
    if not s:
        return None, None, "", ""
    if re.match(r"^\d{5}$", s):
        s = "0" + s
    if re.match(r"^\d{6}$", s):
        month = int(s[:2])
        year = int(s[2:])
        if 1 <= month <= 12:
            return month, year, f"{MONTH_NAMES[month - 1]}-{year}", s
    return parse_gstr1_tax_period(raw_period, financial_year)


def parse_gstr1_tax_period(raw_period: object, financial_year: str = "") -> Tuple[Optional[int], Optional[int], str, str]:
    if raw_period is None:
        return None, None, "", ""
    s = str(raw_period).strip()
    if not s:
        return None, None, "", ""
    if re.match(r"^\d{5,6}$", s):
        return parse_gstr2a_tax_period(s, financial_year)
    clean_key = re.sub(r"[^a-z0-9]", "", s.lower())
    month = MONTH_NAME_MAP.get(clean_key)
    if month:
        month_name = MONTH_NAMES[month - 1]
        start_year, end_year, _ = parse_financial_year(financial_year)
        year = None
        if start_year and end_year:
            year = start_year if month >= 4 else end_year
        display = f"{month_name}-{year}" if year else month_name
        return month, year, display, s
    m = re.match(r"^([A-Za-z]+)[-\s](\d{4})$", s)
    if m:
        m_key = m.group(1).lower()
        if m_key in MONTH_NAME_MAP:
            m_idx = MONTH_NAME_MAP[m_key]
            y = int(m.group(2))
            return m_idx, y, f"{MONTH_NAMES[m_idx - 1]}-{y}", s
    return None, None, s, s


def build_tax_period_range_display(period_objects: List[Dict[str, Any]]) -> str:
    valid = [p for p in period_objects if p.get("month") and p.get("year")]
    if not valid:
        raw_displays = [p.get("display") or p.get("raw") for p in period_objects if p.get("display") or p.get("raw")]
        return raw_displays[0] if raw_displays else ""
    sorted_periods = sorted(valid, key=lambda p: (p["year"], p["month"]))
    first = sorted_periods[0]
    last = sorted_periods[-1]
    if first["year"] == last["year"] and first["month"] == last["month"]:
        return first["display"]
    return f"{first['display']} to {last['display']}"


def _value_to_the_right(grid: List[List[Any]], row_idx: int, start_col: int, max_col: int) -> str:
    if row_idx >= len(grid):
        return ""
    row = grid[row_idx]
    for col in range(start_col + 1, min(max_col + 1, len(row))):
        raw = row[col]
        if raw is None:
            continue
        text = _clean_value(raw)
        if not text:
            continue
        if _match_field(text):
            break
        return text
    return ""


def scan_readme_labels(grid: List[List[Any]], max_row: int = 20, max_col: int = 8) -> Dict[str, str]:
    """Dynamically map metadata fields by scanning label cells in grid."""
    found: Dict[str, str] = {}
    scan_rows = min(max_row, len(grid))

    for r_idx in range(scan_rows):
        row = grid[r_idx]
        scan_cols = min(max_col, len(row))
        for c_idx in range(scan_cols):
            label_cell = row[c_idx]
            field = _match_field(label_cell)
            if not field or field in found:
                continue
            value = _value_to_the_right(grid, r_idx, c_idx, scan_cols)
            if value:
                found[field] = value
    return found


def apply_fallback(
    grid: List[List[Any]],
    found: Dict[str, str],
    fallback: Dict[str, Tuple[int, int]],
) -> Dict[str, str]:
    result = dict(found)
    for field, (row, col) in fallback.items():
        if field in result and result[field]:
            continue
        if row <= 0 or col <= 0:
            continue
        r_idx = row - 1
        c_idx = col - 1
        if r_idx < len(grid) and c_idx < len(grid[r_idx]):
            value = _clean_value(grid[r_idx][c_idx])
            if value:
                result[field] = value
    return result


def fields_to_dealer(fields: Dict[str, str], return_type: str = "gstr2a") -> DealerMetadata:
    _, _, norm_fy = parse_financial_year(fields.get("financial_year", ""))
    raw_period = fields.get("tax_period", "")
    if return_type == "gstr1":
        month, year, disp_period, raw_p = parse_gstr1_tax_period(raw_period, norm_fy)
    else:
        month, year, disp_period, raw_p = parse_gstr2a_tax_period(raw_period, norm_fy)

    return DealerMetadata(
        gstin=fields.get("gstin", "").upper().strip(),
        legal_name=fields.get("legal_name", "").strip(),
        trade_name=fields.get("trade_name", "").strip(),
        financial_year=norm_fy or fields.get("financial_year", "").strip(),
        tax_period=disp_period or raw_period.strip(),
        arn=fields.get("arn", "").strip(),
        arn_date=fields.get("arn_date", "").strip(),
        download_date=fields.get("download_date", "").strip(),
    ).ensure_id()


def extract_from_workbook(wb, return_type: str) -> DealerMetadata:
    readme_name = find_readme_sheet_name(wb.sheetnames)
    if not readme_name:
        return DealerMetadata().ensure_id()

    ws = wb[readme_name]
    raw_grid = [list(r) for r in ws.iter_rows(values_only=True)]
    scanned = scan_readme_labels(raw_grid)
    fallback = GSTR1_FALLBACK if return_type == "gstr1" else GSTR2A_FALLBACK
    merged = apply_fallback(raw_grid, scanned, fallback)
    return fields_to_dealer(merged, return_type)


def extract_from_bytes(content: bytes, return_type: str) -> DealerMetadata:
    wb = openpyxl.load_workbook(io.BytesIO(content), read_only=True, data_only=True)
    try:
        return extract_from_workbook(wb, return_type)
    finally:
        wb.close()


def extract_from_files(
    files: List[Tuple[str, bytes]],
    return_type: str,
) -> List[Tuple[str, DealerMetadata]]:
    if return_type not in {"gstr1", "gstr2a"}:
        raise ValueError(f"Unsupported return type: {return_type}")

    results: List[Tuple[str, DealerMetadata]] = []
    for filename, content in files:
        dealer = extract_from_bytes(content, return_type)
        results.append((filename, dealer))
    return results
